internal/testutils/broker.go

```diff
@@ -0,0 +1,307 @@
+	switch sessionID {
+	case "GAM_invalid":
+		return []map[string]string{
+			{"invalid": "invalid"},
+		}, nil
+
+	case "GAM_empty":
+		return nil, nil
+
+	case "GAM_error":
+		return nil, errors.New("erroring out as requested")
+	}
+
+	return []map[string]string{
+		{"name": "mode1", "label": "Mode 1"},
+	}, nil
+}
+
+// SelectAuthenticationMode returns default values to be used in tests or an error if requested.
+func (b *BrokerMock) SelectAuthenticationMode(ctx context.Context, sessionID, authenticationModeName string) (uiLayoutInfo map[string]string, err error) {
+	switch sessionID {
+	case "SAM_invalid_layout":
+		return map[string]string{
+			"invalid": "invalid",
+		}, nil
+
+	case "SAM_no_layout":
+		return nil, nil
+
+	case "SAM_error":
+		return nil, errors.New("erroring out as requested")
+	}
+
+	return map[string]string{
+		"type":  "form",
+		"label": "Gimme your password",
+		"entry": "chars_password",
+	}, nil
+}
+
+// IsAuthorized returns default values to be used in tests or an error if requested.
+func (b *BrokerMock) IsAuthorized(ctx context.Context, sessionID, authenticationData string) (access, infoUser string, err error) {
+	if sessionID == "IA_error" {
+		return "", "", errors.New("erroring out as requested")
+	}
+
+	// Handles the context that will be assigned for the IsAuthorized handler
+	if _, exists := b.isAuthorizedCalls[sessionID]; exists {
+		return "", "", fmt.Errorf("IsAuthorized already running for session %q", sessionID)
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	b.isAuthorizedCallsMu.Lock()
+	b.isAuthorizedCalls[sessionID] = isAuthorizedCtx{ctx, cancel}
+	b.isAuthorizedCallsMu.Unlock()
+
+	// Cleans the call after it's done
+	defer func() {
+		b.isAuthorizedCallsMu.Lock()
+		delete(b.isAuthorizedCalls, sessionID)
+		b.isAuthorizedCallsMu.Unlock()
+	}()
+
+	access = "allowed"
+	if sessionID == "IA_invalid" {
+		access = "invalid"
+	}
+
+	done := make(chan struct{})
+	go func() {
+		switch sessionID {
+		case "IA_timeout":
+			time.Sleep(time.Second)
+			access = "denied"
+		case "IA_wait":
+			<-ctx.Done()
+			access = "cancelled"
+		}
+		close(done)
+	}()
+	<-done
+
+	infoUser = "{\"some\": \"info\"}"
+	if sessionID == "IA_invalid_info" {
+		infoUser = "invalid"
+	}
+
+	return access, infoUser, nil
+}
+
+// EndSession returns default values to be used in tests or an error if requested.
+func (b *BrokerMock) EndSession(ctx context.Context, sessionID string) (err error) {
+	if sessionID == "ES_error" {
+		return errors.New("erroring out as requested")
+	}
+	return nil
+}
+
+// CancelIsAuthorized cancels an ongoing IsAuthorized call if it exists.
+func (b *BrokerMock) CancelIsAuthorized(ctx context.Context, sessionID string) {
+	b.isAuthorizedCallsMu.Lock()
+	if _, exists := b.isAuthorizedCalls[sessionID]; !exists {
+		b.isAuthorizedCallsMu.Unlock()
+		return
+	}
+	b.isAuthorizedCalls[sessionID].cancelFunc()
+	delete(b.isAuthorizedCalls, sessionID)
+	b.isAuthorizedCallsMu.Unlock()
```

Comment:
you are unlocking in the end of the 2 calls, so I think better to defer Unlock just after the Lock